server/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java

```diff
@@ -178,4 +179,31 @@ public static <T> List<T> parseList(XContentParser parser, CheckedFunction<XCont
         } while (parser.nextToken() != Token.END_ARRAY);
         return list;
     }
+
+    /**
+     * Parses a list of a given type from the given {@code parser} while passing the valueParser the current array index.
+     * Assumes that the parser is currently positioned on a {@link Token#START_ARRAY} token and will fail if it is not.
+     * The returned list may or may not be mutable.
+     *
+     * @param parser      x-content parser
+     * @param valueParser parser for expected list value type
+     * @return list parsed from parser
+     */
+    public static <T> List<T> parseList(XContentParser parser, CheckedBiFunction<XContentParser, Integer, T, IOException> valueParser)
+        throws IOException {
+        XContentParserUtils.ensureExpectedToken(Token.START_ARRAY, parser.currentToken(), parser);
+
+        if (parser.nextToken() == Token.END_ARRAY) {
+            return List.of();
+        }
+
+        final ArrayList<T> list = new ArrayList<>();
+
+        int index = 0;
+        do {
+            list.add(valueParser.apply(parser, index++));
```

Comment:
Same as the old `parseList` but passes in the index of the array as it's parsing. Another idea I had was using a reference within our code to keep track of the current index but this seemed a bit more straight forward.

Reply:
> Same as the old parseList but passes in the index of the array as it's parsing. 

Please state this is in the Javadoc it's hard to see the difference in the 2 methods otherwise